# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/ContentCache.cpp
# Commit: 88cd68e70502
# Full Hash: 88cd68e7050272bf76268c904feef7348d257b56
# Author: Masayuki Nakano <masayuki@d-toybox.com>
# Date: 2022-02-08 17:32:27
# Regressor Bug: 1746104
# File Overlap Count: 1
# Description:
#   Bug 1746104 - part 6-2: Make constructors of `ContentCache::Selection` take `IMENotification::SelectionChangeData` or `WidgetQueryContentEvent` r=m_kato
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D137428
# ==============================================================================

diff -r 2529c39ed6bf -r 88cd68e70502 widget/ContentCache.cpp
--- a/widget/ContentCache.cpp	Mon Feb 07 22:33:39 2022 +0000
+++ b/widget/ContentCache.cpp	Mon Feb 07 22:33:40 2022 +0000
@@ -137,15 +137,7 @@
     return false;
   }
   MOZ_ASSERT(querySelectedTextEvent.mReply->mOffsetAndData.isSome());
-  if (querySelectedTextEvent.mReply->mReversed) {
-    mSelection.emplace(querySelectedTextEvent.mReply->EndOffset(),
-                       querySelectedTextEvent.mReply->StartOffset(),
-                       querySelectedTextEvent.mReply->WritingModeRef());
-  } else {
-    mSelection.emplace(querySelectedTextEvent.mReply->StartOffset(),
-                       querySelectedTextEvent.mReply->EndOffset(),
-                       querySelectedTextEvent.mReply->WritingModeRef());
-  }
+  mSelection.emplace(querySelectedTextEvent);
 
   return CacheCaret(aWidget, aNotification) &&
          CacheTextRects(aWidget, aNotification);
@@ -474,21 +466,16 @@
   return true;
 }
 
-void ContentCacheInChild::SetSelection(nsIWidget* aWidget,
-                                       uint32_t aStartOffset, uint32_t aLength,
-                                       bool aReversed,
-                                       const WritingMode& aWritingMode) {
+void ContentCacheInChild::SetSelection(
+    nsIWidget* aWidget,
+    const IMENotification::SelectionChangeDataBase& aSelectionChangeData) {
   MOZ_LOG(
       sContentCacheLog, LogLevel::Info,
-      ("0x%p SetSelection(aStartOffset=%u, "
-       "aLength=%u, aReversed=%s, aWritingMode=%s), mText=%s",
-       this, aStartOffset, aLength, GetBoolName(aReversed),
-       ToString(aWritingMode).c_str(),
+      ("0x%p SetSelection(aSelectionChangeData=%s), mText=%s", this,
+       ToString(aSelectionChangeData).c_str(),
        PrintStringDetail(mText, PrintStringDetail::kMaxLengthForEditor).get()));
 
-  mSelection = Some(Selection(
-      !aReversed ? aStartOffset : aStartOffset + aLength,
-      !aReversed ? aStartOffset + aLength : aStartOffset, aWritingMode));
+  mSelection = Some(Selection(aSelectionChangeData));
 
   if (mLastCommit.isSome()) {
     // Forget last commit string range if selection is not collapsed
@@ -1659,6 +1646,24 @@
 #endif  // #if MOZ_DIAGNOSTIC_ASSERT_ENABLED
 
 /*****************************************************************************
+ * mozilla::ContentCache::Selection
+ *****************************************************************************/
+
+ContentCache::Selection::Selection(
+    const WidgetQueryContentEvent& aQuerySelectedTextEvent)
+    : mAnchor(UINT32_MAX),
+      mFocus(UINT32_MAX),
+      mWritingMode(aQuerySelectedTextEvent.mReply->WritingModeRef()),
+      mHasRange(aQuerySelectedTextEvent.mReply->mOffsetAndData.isSome()) {
+  MOZ_ASSERT(aQuerySelectedTextEvent.mMessage == eQuerySelectedText);
+  MOZ_ASSERT(aQuerySelectedTextEvent.Succeeded());
+  if (mHasRange) {
+    mAnchor = aQuerySelectedTextEvent.mReply->AnchorOffset();
+    mFocus = aQuerySelectedTextEvent.mReply->FocusOffset();
+  }
+}
+
+/*****************************************************************************
  * mozilla::ContentCache::TextRectArray
  *****************************************************************************/
 